File: backend/src/main/java/com/technicalescaperoom/backend/service/AntiCheatService.java
package com.technicalescaperoom.backend.service;

import com.technicalescaperoom.backend.config.security.PlayerPrincipal;
import com.technicalescaperoom.backend.dto.admin.AntiCheatEventAuditDto;
import com.technicalescaperoom.backend.dto.admin.AntiCheatSummaryDto;
import com.technicalescaperoom.backend.dto.player.AntiCheatEventResponseDto;
import com.technicalescaperoom.backend.dto.player.AntiCheatReportRequest;
import com.technicalescaperoom.backend.entity.*;
import com.technicalescaperoom.backend.enums.AntiCheatViolationType;
import com.technicalescaperoom.backend.enums.EventStatus;
import com.technicalescaperoom.backend.enums.GameEventType;
import com.technicalescaperoom.backend.enums.TeamGameState;
import com.technicalescaperoom.backend.exception.EventUnavailableException;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.*;
import com.technicalescaperoom.backend.service.admin.LeaderboardService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
@Service
@RequiredArgsConstructor
public class AntiCheatService {

    private final AntiCheatEventRepository antiCheatEventRepository;
    private final TeamAntiCheatSummaryRepository teamAntiCheatSummaryRepository;
    private final TeamRepository teamRepository;
    private final PlayerRepository playerRepository;
    private final AuditService auditService;
    private final GameWebSocketPublisher webSocketPublisher;
    private final LeaderboardService leaderboardService;

    @Value("${codexcape.anticheat.tab-switch-penalty:10}")
    private int tabSwitchPenalty = 10;

    @Value("${codexcape.anticheat.fullscreen-exit-penalty:15}")
    private int fullscreenExitPenalty = 15;

    @Value("${codexcape.anticheat.prolonged-hidden-penalty:15}")
    private int prolongedHiddenPenalty = 15;

    @Value("${codexcape.anticheat.cooldown-seconds:10}")
    private int cooldownSeconds = 10;

    @Value("${codexcape.anticheat.correlation-window-seconds:5}")
    private int correlationWindowSeconds = 5;

    @Value("${codexcape.anticheat.prolonged-threshold-seconds:30}")
    private int prolongedThresholdSeconds = 30;

    // Track when player tab was hidden to compute duration and detect prolonged absence
    private final Map<Long, Instant> playerHiddenStartTime = new ConcurrentHashMap<>();
    // Track recent violations for high-speed deduplication
    private final Map<String, Instant> recentViolationCache = new ConcurrentHashMap<>();

    @Transactional
    public AntiCheatEventResponseDto processPlayerEvent(PlayerPrincipal principal, AntiCheatReportRequest request) {
        if (principal == null) {
            throw new ResourceNotFoundException("No authenticated player session found.");
        }

        Team team = teamRepository.findById(principal.getTeamId())
                .orElseThrow(() -> new ResourceNotFoundException("Team not found for ID: " + principal.getTeamId()));

        Player player = playerRepository.findById(principal.getPlayerId())
                .orElseThrow(() -> new ResourceNotFoundException("Player not found for ID: " + principal.getPlayerId()));

        if (!player.getTeam().getId().equals(team.getId())) {
            throw new ResourceNotFoundException("Player does not belong to specified team.");
        }

        Event event = team.getEvent();

        // 1. ACTIVE GAME ONLY: Only actively penalize during competitive IN_PROGRESS gameplay
        if (team.getGameState() != TeamGameState.IN_PROGRESS) {
            log.debug("Anti-cheat event ignored: Team {} is in {} state (only IN_PROGRESS is penalized).",
                    team.getTeamCode(), team.getGameState());
            return AntiCheatEventResponseDto.builder()
                    .accepted(false)
                    .deduplicated(false)
                    .message("Event ignored: Game is not actively in progress.")
                    .build();
        }

        if (event.getStatus() != EventStatus.RUNNING && event.getStatus() != EventStatus.READY) {
            log.debug("Anti-cheat event ignored: Event {} status is {}.", event.getName(), event.getStatus());
            return AntiCheatEventResponseDto.builder()
                    .accepted(false)
                    .deduplicated(false)
                    .message("Event ignored: Event is not running.")
                    .build();
        }

        String rawType = (request.getEventType() != null) ? request.getEventType().trim().toUpperCase() : "";
        Instant now = Instant.now();

        // 2. Handle recovery event (VISIBILITY_RESTORED / FOCUS_RESTORED)
        if ("VISIBILITY_RESTORED".equals(rawType) || "PAGE_VISIBLE".equals(rawType) || "FOCUS_RESTORED".equals(rawType)) {
            Instant hiddenSince = playerHiddenStartTime.remove(player.getId());
            if (hiddenSince != null) {
                long awaySeconds = Duration.between(hiddenSince, now).getSeconds();
                if (awaySeconds >= prolongedThresholdSeconds) {
                    log.warn("Player {} on Team {} was away for {}s (prolonged absence threshold: {}s).",
                            player.getDisplayName(), team.getTeamCode(), awaySeconds, prolongedThresholdSeconds);
                    return recordViolation(team, player, event, AntiCheatViolationType.PROLONGED_PAGE_HIDDEN,
                            awaySeconds * 1000L, now, "Prolonged background absence for " + awaySeconds + " seconds.");
                }
            }
            return AntiCheatEventResponseDto.builder()
                    .accepted(false)
                    .deduplicated(false)
                    .message("Visibility restored recorded. No penalty applied.")
                    .build();
        }

        // 3. Map raw event type to authoritative ViolationType
        AntiCheatViolationType violationType;
        if ("TAB_SWITCH".equals(rawType) || "PAGE_HIDDEN".equals(rawType) || "VISIBILITY_HIDDEN".equals(rawType)) {
            violationType = AntiCheatViolationType.TAB_SWITCH;
            playerHiddenStartTime.putIfAbsent(player.getId(), now);
        } else if ("FULLSCREEN_EXIT".equals(rawType) || "FULLSCREEN_CHANGE".equals(rawType)) {
            violationType = AntiCheatViolationType.FULLSCREEN_EXIT;
        } else if ("PROLONGED_PAGE_HIDDEN".equals(rawType)) {
            violationType = AntiCheatViolationType.PROLONGED_PAGE_HIDDEN;
        } else {
            log.debug("Ignored unmonitored anti-cheat event type '{}'", rawType);
            return AntiCheatEventResponseDto.builder()
                    .accepted(false)
                    .deduplicated(false)
                    .message("Ignored unrecognized event type.")
                    .build();
        }

        // 4. Server-Side Incident Deduplication & Correlation
        // A. Correlated incident: If player exited fullscreen and visibility hidden within correlation window (5s), correlate to 1 incident
        String correlationKey = "recent_fullscreen_" + player.getId();
        Instant lastFullscreen = recentViolationCache.get(correlationKey);
        if (violationType == AntiCheatViolationType.TAB_SWITCH && lastFullscreen != null) {
            if (Duration.between(lastFullscreen, now).getSeconds() < correlationWindowSeconds) {
                log.info("Correlating TAB_SWITCH with recent FULLSCREEN_EXIT for Player {} on Team {}. Deduplicated.",
                        player.getDisplayName(), team.getTeamCode());
                return AntiCheatEventResponseDto.builder()
                        .accepted(false)
                        .deduplicated(true)
                        .message("Correlated with recent fullscreen exit. Deduplicated.")
                        .build();
            }
        }

        // B. Cooldown deduplication: Do not issue repeated penalties of the same type within cooldownSeconds
        String cooldownKey = "cooldown_" + player.getId() + "_" + violationType.name();
        Instant lastViolation = recentViolationCache.get(cooldownKey);
        if (lastViolation != null && Duration.between(lastViolation, now).getSeconds() < cooldownSeconds) {
            log.info("Deduplicating rapid {} event for Player {} (cooldown: {}s).",
                    violationType, player.getDisplayName(), cooldownSeconds);
            return AntiCheatEventResponseDto.builder()
                    .accepted(false)
                    .deduplicated(true)
                    .violationType(violationType)
                    .message("Duplicate event within cooldown window. Deduplicated.")
                    .build();
        }

        // Record cooldown marker
        recentViolationCache.put(cooldownKey, now);
        if (violationType == AntiCheatViolationType.FULLSCREEN_EXIT) {
            recentViolationCache.put(correlationKey, now);
        }

        // 5. Authoritative Penalty Calculation & Recording
        return recordViolation(team, player, event, violationType, null, now, request.getMetadata());
    }

    private AntiCheatEventResponseDto recordViolation(Team team, Player player, Event event,
                                                      AntiCheatViolationType violationType,
                                                      Long durationMs, Instant detectedAt, String metadata) {
        int penaltyPoints = getPenaltyPointsForType(violationType);
        String incidentKey = "inc_" + team.getId() + "_p" + player.getPlayerNumber() + "_" + violationType.name() + "_" + (detectedAt.toEpochMilli() / (cooldownSeconds * 1000L));

        AntiCheatEvent auditRecord = AntiCheatEvent.builder()
                .team(team)
                .player(player)
                .event(event)
                .violationType(violationType)
                .detectedAt(detectedAt)
                .durationMs(durationMs)
                .penaltyPoints(penaltyPoints)
                .incidentKey(incidentKey)
                .metadata(metadata)
                .build();
        antiCheatEventRepository.save(auditRecord);

        // Update or create Team summary atomically
        TeamAntiCheatSummary summary = teamAntiCheatSummaryRepository.findByTeamId(team.getId())
                .orElseGet(() -> TeamAntiCheatSummary.builder()
                        .team(team)
                        .totalPenaltyPoints(0)
                        .totalViolations(0)
                        .tabSwitchCount(0)
                        .fullscreenExitCount(0)
                        .prolongedHiddenCount(0)
                        .build());

        summary.setTotalPenaltyPoints(summary.getTotalPenaltyPoints() + penaltyPoints);
        summary.setTotalViolations(summary.getTotalViolations() + 1);
        summary.setLastViolationAt(detectedAt);

        if (violationType == AntiCheatViolationType.TAB_SWITCH) {
            summary.setTabSwitchCount(summary.getTabSwitchCount() + 1);
        } else if (violationType == AntiCheatViolationType.FULLSCREEN_EXIT) {
            summary.setFullscreenExitCount(summary.getFullscreenExitCount() + 1);
        } else if (violationType == AntiCheatViolationType.PROLONGED_PAGE_HIDDEN) {
            summary.setProlongedHiddenCount(summary.getProlongedHiddenCount() + 1);
        }

        teamAntiCheatSummaryRepository.saveAndFlush(summary);

        log.warn("🚨 ANTI-CHEAT VIOLATION: Team {} | Player {} (P{}) | Type: {} | -{} pts | Team Total: -{} pts",
                team.getTeamCode(), player.getDisplayName(), player.getPlayerNumber(),
                violationType, penaltyPoints, summary.getTotalPenaltyPoints());

        // Audit Trail Log
        auditService.logEvent(
                GameEventType.ANTI_CHEAT_VIOLATION,
                event,
                team,
                player,
                String.format("{\"violationType\":\"%s\",\"penaltyPoints\":%d,\"teamTotal\":%d}",
                        violationType.name(), penaltyPoints, summary.getTotalPenaltyPoints()),
                "ANTI_CHEAT"
        );

        String violationLabel = formatViolationLabel(violationType);
        String alertMsg = String.format("ANTI-CHEAT ALERT: %s DETECTED (Player %d). Team Penalty: -%d pts (Team Total: -%d pts).",
                violationLabel, player.getPlayerNumber(), penaltyPoints, summary.getTotalPenaltyPoints());

        // Realtime WebSockets:
        // 1. Notify Team channel (/topic/team/{id}) with alert & team total
        webSocketPublisher.notifyTeamAntiCheatAlert(
                team.getId(),
                team.getTeamCode(),
                player.getPlayerNumber(),
                violationType.name(),
                penaltyPoints,
                summary.getTotalPenaltyPoints(),
                alertMsg
        );

        // 2. Notify Admin channel (/topic/admin) with full forensic event
        webSocketPublisher.notifyAdminAntiCheatEvent(
                team.getId(),
                team.getTeamCode(),
                player.getId(),
                player.getPlayerNumber(),
                player.getDisplayName(),
                violationType.name(),
                penaltyPoints,
                summary.getTotalPenaltyPoints(),
                summary.getTotalViolations(),
                alertMsg
        );

        // 3. Recalculate Leaderboard Ranks immediately
        leaderboardService.recalculateAndBroadcastRanks(event.getId(), webSocketPublisher);

        return AntiCheatEventResponseDto.builder()
                .accepted(true)
                .deduplicated(false)
                .incidentId(auditRecord.getId())
                .violationType(violationType)
                .penaltyPoints(penaltyPoints)
                .teamTotalPenalties(summary.getTotalPenaltyPoints())
                .message(alertMsg)
                .build();
    }

    private int getPenaltyPointsForType(AntiCheatViolationType type) {
        return switch (type) {
            case TAB_SWITCH -> tabSwitchPenalty;
            case FULLSCREEN_EXIT -> fullscreenExitPenalty;
            case PROLONGED_PAGE_HIDDEN -> prolongedHiddenPenalty;
        };
    }

    private String formatViolationLabel(AntiCheatViolationType type) {
        return switch (type) {
            case TAB_SWITCH -> "TAB SWITCH";
            case FULLSCREEN_EXIT -> "FULLSCREEN EXIT";
            case PROLONGED_PAGE_HIDDEN -> "PROLONGED INACTIVITY";
        };
    }

    @Transactional(readOnly = true)
    public AntiCheatSummaryDto getTeamSummary(Long teamId) {
        Team team = teamRepository.findById(teamId)
                .orElseThrow(() -> new ResourceNotFoundException("Team not found"));
        TeamAntiCheatSummary summary = teamAntiCheatSummaryRepository.findByTeamId(teamId)
                .orElse(null);

        return AntiCheatSummaryDto.builder()
                .teamId(team.getId())
                .teamCode(team.getTeamCode())
                .teamName(team.getTeamName())
                .totalPenaltyPoints(summary != null ? summary.getTotalPenaltyPoints() : 0)
                .totalViolations(summary != null ? summary.getTotalViolations() : 0)
                .tabSwitchCount(summary != null ? summary.getTabSwitchCount() : 0)
                .fullscreenExitCount(summary != null ? summary.getFullscreenExitCount() : 0)
                .prolongedHiddenCount(summary != null ? summary.getProlongedHiddenCount() : 0)
                .lastViolationAt(summary != null ? summary.getLastViolationAt() : null)
                .build();
    }

    @Transactional(readOnly = true)
    public List<AntiCheatSummaryDto> getEventSummaries(Long eventId) {
        List<Team> teams = teamRepository.findByEventId(eventId);
        if (teams.isEmpty()) return Collections.emptyList();

        List<Long> teamIds = teams.stream().map(Team::getId).toList();
        Map<Long, TeamAntiCheatSummary> summaryMap = teamAntiCheatSummaryRepository.findByTeamIdIn(teamIds).stream()
                .collect(java.util.stream.Collectors.toMap(TeamAntiCheatSummary::getTeamId, s -> s));

        List<AntiCheatSummaryDto> dtos = new ArrayList<>();
        for (Team team : teams) {
            TeamAntiCheatSummary s = summaryMap.get(team.getId());
            dtos.add(AntiCheatSummaryDto.builder()
                    .teamId(team.getId())
                    .teamCode(team.getTeamCode())
                    .teamName(team.getTeamName())
                    .totalPenaltyPoints(s != null ? s.getTotalPenaltyPoints() : 0)
                    .totalViolations(s != null ? s.getTotalViolations() : 0)
                    .tabSwitchCount(s != null ? s.getTabSwitchCount() : 0)
                    .fullscreenExitCount(s != null ? s.getFullscreenExitCount() : 0)
                    .prolongedHiddenCount(s != null ? s.getProlongedHiddenCount() : 0)
                    .lastViolationAt(s != null ? s.getLastViolationAt() : null)
                    .build());
        }
        dtos.sort((a, b) -> Integer.compare(b.getTotalPenaltyPoints(), a.getTotalPenaltyPoints()));
        return dtos;
    }

    @Transactional(readOnly = true)
    public List<AntiCheatEventAuditDto> getEventViolations(Long eventId) {
        List<AntiCheatEvent> events = antiCheatEventRepository.findByEventIdOrderByDetectedAtDesc(eventId);
        return events.stream().map(this::toAuditDto).toList();
    }

    @Transactional(readOnly = true)
    public List<AntiCheatEventAuditDto> getTeamViolations(Long teamId) {
        List<AntiCheatEvent> events = antiCheatEventRepository.findByTeamIdOrderByDetectedAtDesc(teamId);
        return events.stream().map(this::toAuditDto).toList();
    }

    private AntiCheatEventAuditDto toAuditDto(AntiCheatEvent e) {
        return AntiCheatEventAuditDto.builder()
                .id(e.getId())
                .teamId(e.getTeam().getId())
                .teamCode(e.getTeam().getTeamCode())
                .teamName(e.getTeam().getTeamName())
                .playerId(e.getPlayer().getId())
                .playerName(e.getPlayer().getDisplayName())
                .playerNumber(e.getPlayer().getPlayerNumber())
                .violationType(e.getViolationType().name())
                .penaltyPoints(e.getPenaltyPoints())
                .durationMs(e.getDurationMs())
                .detectedAt(e.getDetectedAt())
                .metadata(e.getMetadata())
                .build();
    }
}

File: backend/services/antiCheatService.js
const antiCheatEventRepository = require('../repositories/antiCheatEventRepository');
const teamAntiCheatSummaryRepository = require('../repositories/teamAntiCheatSummaryRepository');
const teamRepository = require('../repositories/teamRepository');
const playerRepository = require('../repositories/playerRepository');
const eventRepository = require('../repositories/eventRepository');
const auditService = require('./auditService');
const webSocketPublisher = require('./webSocketService');
const scoringService = require('./scoringService');
const leaderboardService = require('./admin/leaderboardService');

class AntiCheatService {
  constructor() {
    this.tabSwitchPenalty = 10;
    this.fullscreenExitPenalty = 15;
    this.prolongedHiddenPenalty = 15;
    this.cooldownSeconds = 10;
    this.correlationWindowSeconds = 5;
    this.prolongedThresholdSeconds = 30;

    this.playerHiddenStartTime = new Map();
    this.recentViolationCache = new Map();
  }

  async processPlayerEvent(principal, request) {
    if (!principal) {
      const err = new Error('No authenticated player session found.');
      err.status = 404;
      throw err;
    }

    const team = await teamRepository.findById(principal.teamId);
    if (!team) {
      const err = new Error(`Team not found for ID: ${principal.teamId}`);
      err.status = 404;
      throw err;
    }

    const player = await playerRepository.findById(principal.playerId);
    if (!player || player.teamId !== team.id) {
      const err = new Error('Player does not belong to specified team.');
      err.status = 404;
      throw err;
    }

    const event = await eventRepository.findById(team.eventId);

    // 1. ACTIVE GAME ONLY: Only actively penalize during competitive IN_PROGRESS gameplay
    if (team.gameState !== 'IN_PROGRESS') {
      return {
        accepted: false,
        deduplicated: false,
        message: 'Event ignored: Game is not actively in progress.'
      };
    }

    if (!event || (event.status !== 'RUNNING' && event.status !== 'READY')) {
      return {
        accepted: false,
        deduplicated: false,
        message: 'Event ignored: Event is not running.'
      };
    }

    const rawType = request.eventType ? String(request.eventType).trim().toUpperCase() : '';
    const now = Date.now();

    // 2. Handle recovery event (VISIBILITY_RESTORED / FOCUS_RESTORED)
    if (rawType === 'VISIBILITY_RESTORED' || rawType === 'PAGE_VISIBLE' || rawType === 'FOCUS_RESTORED') {
      const hiddenSince = this.playerHiddenStartTime.get(player.id);
      this.playerHiddenStartTime.delete(player.id);

      if (hiddenSince) {
        const awaySeconds = Math.floor((now - hiddenSince) / 1000);
        if (awaySeconds >= this.prolongedThresholdSeconds) {
          return await this._recordViolation(
            team,
            player,
            event,
            'PROLONGED_PAGE_HIDDEN',
            awaySeconds * 1000,
            new Date(now).toISOString(),
            `Prolonged background absence for ${awaySeconds} seconds.`
          );
        }
      }
      return {
        accepted: false,
        deduplicated: false,
        message: 'Visibility restored recorded. No penalty applied.'
      };
    }

    // 3. Map raw event type to authoritative ViolationType
    let violationType = null;
    if (rawType === 'TAB_SWITCH' || rawType === 'PAGE_HIDDEN' || rawType === 'VISIBILITY_HIDDEN') {
      violationType = 'TAB_SWITCH';
      if (!this.playerHiddenStartTime.has(player.id)) {
        this.playerHiddenStartTime.set(player.id, now);
      }
    } else if (rawType === 'FULLSCREEN_EXIT' || rawType === 'FULLSCREEN_CHANGE') {
      violationType = 'FULLSCREEN_EXIT';
    } else if (rawType === 'PROLONGED_PAGE_HIDDEN') {
      violationType = 'PROLONGED_PAGE_HIDDEN';
    } else {
      return {
        accepted: false,
        deduplicated: false,
        message: 'Ignored unrecognized event type.'
      };
    }

    // 4. Server-Side Incident Deduplication & Correlation
    // A. Correlated incident: If player exited fullscreen and visibility hidden within correlation window (5s)
    const correlationKey = `recent_fullscreen_${player.id}`;
    const lastFullscreen = this.recentViolationCache.get(correlationKey);
    if (violationType === 'TAB_SWITCH' && lastFullscreen) {
      if ((now - lastFullscreen) / 1000 < this.correlationWindowSeconds) {
        return {
          accepted: false,
          deduplicated: true,
          message: 'Correlated with recent fullscreen exit. Deduplicated.'
        };
      }
    }

    // B. Cooldown deduplication: Do not issue repeated penalties of the same type within cooldownSeconds
    const cooldownKey = `cooldown_${player.id}_${violationType}`;
    const lastViolation = this.recentViolationCache.get(cooldownKey);
    if (lastViolation && (now - lastViolation) / 1000 < this.cooldownSeconds) {
      return {
        accepted: false,
        deduplicated: true,
        violationType,
        message: 'Duplicate event within cooldown window. Deduplicated.'
      };
    }

    // Record cooldown marker
    this.recentViolationCache.set(cooldownKey, now);
    if (violationType === 'FULLSCREEN_EXIT') {
      this.recentViolationCache.set(correlationKey, now);
    }

    // 5. Authoritative Penalty Calculation & Recording
    return await this._recordViolation(
      team,
      player,
      event,
      violationType,
      0,
      new Date(now).toISOString(),
      request.metadata || null
    );
  }

  async _recordViolation(team, player, event, violationType, durationMs, detectedAt, metadata) {
    let summary = await teamAntiCheatSummaryRepository.findByTeamId(team.id);
    if (!summary) {
      summary = {
        teamId: team.id,
        totalPenaltyPoints: 0,
        totalViolations: 0,
        tabSwitchCount: 0,
        fullscreenExitCount: 0,
        prolongedHiddenCount: 0
      };
    }

    let incidentNumber;
    if (violationType === 'TAB_SWITCH' || violationType === 'PROLONGED_PAGE_HIDDEN') {
      incidentNumber = summary.tabSwitchCount + 1;
    } else {
      incidentNumber = summary.fullscreenExitCount + 1;
    }

    const incidentKey = `inc_${team.id}_p${player.playerNumber}_${violationType}_#${incidentNumber}`;

    // Anti-cheat events do NOT reduce game score directly in standard config
    const penaltyPoints = 0;
    await scoringService.recordAntiCheatPenalty(team.id, player.id, violationType, penaltyPoints, incidentKey);

    const sanitizedMetadata = metadata ? String(metadata).substring(0, 255) : null;

    const auditRecord = await antiCheatEventRepository.recordEvent({
      teamId: team.id,
      playerId: player.id,
      eventId: event.id,
      violationType,
      detectedAt,
      durationMs: durationMs || 0,
      penaltyPoints,
      incidentKey,
      metadata: sanitizedMetadata
    });

    summary.totalPenaltyPoints = 0;
    summary.totalViolations = (summary.totalViolations || 0) + 1;
    summary.lastViolationAt = detectedAt;

    if (violationType === 'TAB_SWITCH') {
      summary.tabSwitchCount = (summary.tabSwitchCount || 0) + 1;
    } else if (violationType === 'FULLSCREEN_EXIT') {
      summary.fullscreenExitCount = (summary.fullscreenExitCount || 0) + 1;
    } else if (violationType === 'PROLONGED_PAGE_HIDDEN') {
      summary.prolongedHiddenCount = (summary.prolongedHiddenCount || 0) + 1;
    }

    await teamAntiCheatSummaryRepository.upsert(summary);

    // Audit trail log
    await auditService.logEvent(
      'ANTI_CHEAT_VIOLATION',
      event,
      team,
      player,
      JSON.stringify({ violationType, totalViolations: summary.totalViolations }),
      'ANTI_CHEAT'
    );

    const violationLabel = this._formatViolationLabel(violationType);
    const alertMsg = `ANTI-CHEAT ALERT: ${violationLabel} DETECTED (Player ${player.playerNumber}). Incident logged.`;

    // Realtime WebSockets:
    webSocketPublisher.notifyTeamAntiCheatAlert(
      team.id,
      team.teamCode,
      player.playerNumber,
      violationType,
      0,
      0,
      alertMsg
    );

    webSocketPublisher.notifyAdminAntiCheatEvent(
      team.id,
      team.teamCode,
      player.id,
      player.playerNumber,
      player.displayName,
      violationType,
      0,
      0,
      summary.totalViolations,
      alertMsg
    );

    // Recalculate ranks
    leaderboardService.recalculateAndBroadcastRanks(event.id);

    return {
      accepted: true,
      deduplicated: false,
      incidentId: auditRecord.id,
      violationType,
      penaltyPoints: 0,
      teamTotalPenalties: 0,
      message: alertMsg
    };
  }

  _formatViolationLabel(type) {
    switch (type) {
      case 'TAB_SWITCH':
        return 'TAB SWITCH';
      case 'FULLSCREEN_EXIT':
        return 'FULLSCREEN EXIT';
      case 'PROLONGED_PAGE_HIDDEN':
        return 'PROLONGED INACTIVITY';
      default:
        return type;
    }
  }

  async getTeamSummary(teamId) {
    const team = await teamRepository.findById(teamId);
    if (!team) {
      const err = new Error('Team not found');
      err.status = 404;
      throw err;
    }

    const summary = await teamAntiCheatSummaryRepository.findByTeamId(teamId);

    return {
      teamId: team.id,
      teamCode: team.teamCode,
      teamName: team.teamName,
      totalPenaltyPoints: summary ? summary.totalPenaltyPoints : 0,
      totalViolations: summary ? summary.totalViolations : 0,
      tabSwitchCount: summary ? summary.tabSwitchCount : 0,
      fullscreenExitCount: summary ? summary.fullscreenExitCount : 0,
      prolongedHiddenCount: summary ? summary.prolongedHiddenCount : 0,
      lastViolationAt: summary ? summary.lastViolationAt : null
    };
  }

  async getEventSummaries(eventId) {
    const teams = await teamRepository.findByEventId(eventId);
    if (!teams || teams.length === 0) return [];

    const summaries = await teamAntiCheatSummaryRepository.findByEventId(eventId);
    const summaryMap = new Map();
    for (const s of summaries) {
      summaryMap.set(s.teamId, s);
    }

    const dtos = [];
    for (const team of teams) {
      const s = summaryMap.get(team.id);
      dtos.push({
        teamId: team.id,
        teamCode: team.teamCode,
        teamName: team.teamName,
        totalPenaltyPoints: s ? s.totalPenaltyPoints : 0,
        totalViolations: s ? s.totalViolations : 0,
        tabSwitchCount: s ? s.tabSwitchCount : 0,
        fullscreenExitCount: s ? s.fullscreenExitCount : 0,
        prolongedHiddenCount: s ? s.prolongedHiddenCount : 0,
        lastViolationAt: s ? s.lastViolationAt : null
      });
    }

    dtos.sort((a, b) => b.totalPenaltyPoints - a.totalPenaltyPoints);
    return dtos;
  }

  async getEventViolations(eventId) {
    return await antiCheatEventRepository.findByEventId(eventId);
  }

  async getTeamViolations(teamId) {
    return await antiCheatEventRepository.findByTeamId(teamId);
  }

  resetDeduplicationCache() {
    this.playerHiddenStartTime.clear();
    this.recentViolationCache.clear();
  }
}

module.exports = new AntiCheatService();
